### PR TITLE
Enable NopEliminatoin inside MO Backend

### PR DESCRIPTION
**Pull request by GlebKazantaev**
### Description
This  PR includes some dependent changes related to `AlgebraicSimplification` and `NopElimination` refactoring:
* `AlgebraicSimplification` pass was refactored and finally removed. `SimplifyGatherShapeOf` passes were moved into `SimplifyShapeOfSubGraph`.
* `AlgebraicSimplification` tests were either moved to `NopElimination` tests or removed because the doesn't test enything.
* `NopElimination` pass was refactored. `EliminateConvert`, `EliminateConvertNonZero`, `EliminateConcat` was updated to use `MatcherPass`.
* `NopElimination` was parametrized to work in shape independent mode.
* Implemented `EliminateSplit` and `EliminateTranspose` passes.
* `EliminatePad` was moved to `NopElimination`.
* Added `NopElimination` pass to the MOC transformations pipeline.

GraphRewrite facilities was updated to use `weak_ptr` for intermediate node storage. This fix the problem that after replacement even if node has no consumers it wasn't removed. So for some cases (eliminations) it leads to the wrong number of operations consumers so some passes wasn't applied.

Validation on DL Benchmark shows that useless Transposes were removed in some models.

### nGraph Transformations Transition Status
<details>
  <summary>Click to expand!</summary>

|Final MOC Pipeline|Enabled|
|---|:---:|
|InitNodeInfo|:white_check_mark:|
|ConstantFolding| |
|SimplifyShapeOfSubGraph|:white_check_mark:|
|RemoveFilteringBoxesBySize|:white_check_mark:|
|ConvertQuantizeDequantize|:white_check_mark:|
| | |
|TransposeFQReduction|:white_check_mark:|
|TransposeReduction|:white_check_mark:|
|TransposeFuse|:white_check_mark:|
|SplitSqueezeConcatFusion|:white_check_mark:|
| | |
|EliminateUnsqueezeGather|:white_check_mark:|
|**AlgebraicSimplification**| removed |
|**NopElimination**|:tada: |
| | |
|ConvertScatterElementsToScatter|:white_check_mark:|
|BroadcastElementwiseFusion|:white_check_mark:|
|SoftPlusFusion|:white_check_mark:|
|SoftPlusToMishFusion|:white_check_mark:|
|SwishFusion|:white_check_mark:|
|HSwishFusion|:white_check_mark:|
|HSigmoidFusion|:white_check_mark:|
|NormalizeL2Fusion| |
|ClampFusion|:white_check_mark:|
|PadFusion|:white_check_mark:|
|SoftmaxFusion| |
|MVNFusion|:white_check_mark:| 
|DilatedConvolutionConverter|:white_check_mark:|
|GeluFusion|:white_check_mark:|
| | |
|BinarizeWeights|:white_check_mark:|
|ConvToBinaryConv|:white_check_mark:|
| | |
|BatchNormDecomposition|:white_check_mark: | 
| | |
|LinOpSequenceFusion|:white_check_mark: |
| | |
|ConvolutionMultiplyFusion|:white_check_mark: |
|GroupConvolutionMultiplyFusion|:white_check_mark: |
|ConvolutionBackpropDataMultiplyFusion|:white_check_mark: |
|GroupConvolutionBackpropDataMultiplyFusion|:white_check_mark: |

_Note: transformations marked as **bold** and with :tada: were enabled in current PR_
</details>

### CODE REVIEW

File: inference-engine/src/transformations/src/transformations/common_optimizations/nop_elimination.cpp [hunk 1]
```diff
@@ -336,20 +358,168 @@ NAME() { \
...
+    };
+
+    auto m = std::make_shared<pattern::Matcher>(convert_pattern, matcher_name);
+    this->register_matcher(m, callback);
+}
+
+NGRAPH_RTTI_DEFINITION(pass::EliminateSplit, "EliminateSplit", 0);
+
+pass::EliminateSplit::EliminateSplit() {
+    MATCHER_SCOPE(EliminateConcat);
```
Comment (jane-intel) [inline inference-engine/src/transformations/src/transformations/common_optimizations/nop_elimination.cpp:462]:
```suggestion
    MATCHER_SCOPE(EliminateSplit);
```

File: inference-engine/src/transformations/src/transformations/common_optimizations/simplify_shape_of_sub_graph.cpp [hunk 2]
```diff
@@ -121,16 +122,85 @@ ngraph::pass::GatherNopElimination::GatherNopElimination() {
...
+    MATCHER_SCOPE(SimplifyGatherShapeOf);
+    const auto gather_pattern = ngraph::pattern::wrap_type<op::util::GatherBase>();
+    const auto shape_of_pattern = ngraph::pattern::wrap_type<opset2::ShapeOf, opset3::ShapeOf>({gather_pattern});
+
+    ngraph::matcher_pass_callback callback = [](pattern::Matcher& m) {
+        auto node = m.get_match_root();
+        auto gather = as_type_ptr<opset3::Gather>(node->input_value(0).get_node_shared_ptr());
+        if (!gather) {
+            return false;
+        }
```
Comment (jane-intel) [inline inference-engine/src/transformations/src/transformations/common_optimizations/simplify_shape_of_sub_graph.cpp:137]: we match gather base but require opset3 version.

**Review by jane-intel — COMMENTED** [review]
there are several nitpicks

File: inference-engine/src/transformations/src/transformations/common_optimizations/simplify_shape_of_sub_graph.cpp (hunk 2 above)
Comment (GlebKazantaev) [reply inference-engine/src/transformations/src/transformations/common_optimizations/simplify_shape_of_sub_graph.cpp:137]: Discussed offline that this transformation needs to be updated to support new Gather versions (opset7, opset8). The ticket in transformations backlog will be created.

File: inference-engine/src/transformations/src/transformations/common_optimizations/nop_elimination.cpp (hunk 1 above)
Comment (GlebKazantaev) [reply inference-engine/src/transformations/src/transformations/common_optimizations/nop_elimination.cpp:462]: If you don't mind I will fix it within https://github.com/openvinotoolkit/openvino/pull/6916 PR.

**Review by jane-intel — APPROVED** [review]

File: inference-engine/src/transformations/src/transformations/common_optimizations/simplify_shape_of_sub_graph.cpp (hunk 2 above)
Comment (GlebKazantaev) [reply inference-engine/src/transformations/src/transformations/common_optimizations/simplify_shape_of_sub_graph.cpp:137]: Ticket: XXX-61783
